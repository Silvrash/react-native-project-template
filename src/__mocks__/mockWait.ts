export async function mockWait(timeout: number) {
	return new Promise(resolve => {
		setTimeout(() => {
			resolve(true);
		}, timeout);
	});
}
