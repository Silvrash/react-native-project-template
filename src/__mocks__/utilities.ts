import { useNavigation } from '@react-navigation/native';

export function setupNavigationMock() {
	const navigate = jest.fn();
	const goBack = jest.fn();
	const dispatch = jest.fn();
	const removeListener = jest.fn();
	const addListener = jest.fn();

	(useNavigation as jest.Mock).mockReturnValue({
		navigate,
		goBack,
		dispatch,
		addListener,
		removeListener,
	});
	addListener.mockImplementation((_key: string, fn: (e: any) => void) => {
		fn({ preventDefault: () => null });
	});
}
