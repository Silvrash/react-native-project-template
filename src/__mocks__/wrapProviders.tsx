import { render, RenderOptions } from '@testing-library/react-native';
import { PropsWithChildren } from 'react';
import NavigationWrapper from 'src/wrapper.navigation';
import ThemeWrapper from '../wrapper.theme';

const AllTheProviders = ({ children }: PropsWithChildren) => {
	return (
		<ThemeWrapper>
			<NavigationWrapper>{children}</NavigationWrapper>
		</ThemeWrapper>
	);
};

export function renderWithProviders<T>(ui: React.ReactElement<T>, options?: RenderOptions) {
	return render(ui, {
		wrapper: ({ children }) => <AllTheProviders>{children}</AllTheProviders>,
		...options,
	});
}
