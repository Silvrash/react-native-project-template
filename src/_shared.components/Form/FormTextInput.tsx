import { IInputProps, IconButton, Input } from 'native-base';
import React, { useEffect, useState } from 'react';
import EyeWithLineIcon from 'src/_shared.assets/eye-with-line.svg';
import EyeIcon from 'src/_shared.assets/eye.svg';
import { IconType } from 'src/_shared/types';
import ConditionRender from '../ConditionRender';
import FormBase, { FormBaseProps } from './FormBase';

export interface FormTextInputProps extends IInputProps, Omit<FormBaseProps, 'input'> {
	leftIcon?: IconType;
	rightIcon?: IconType;
}

const FormTextInput: React.FC<FormTextInputProps> = ({
	label,
	isRequired,
	error,
	helperText,
	rootStyle,
	labelColor,
	leftIcon,
	rightIcon,
	...textInputProps
}) => {
	const [secureTextEntry, setSecureTextEntry] = useState(!!textInputProps.secureTextEntry);

	useEffect(
		() => setSecureTextEntry(!!textInputProps.secureTextEntry),
		[textInputProps.secureTextEntry],
	);

	function toggleSecureTextEntry() {
		setSecureTextEntry(value => !value);
	}

	return (
		<FormBase
			label={label}
			isRequired={isRequired}
			error={error}
			helperText={helperText}
			rootStyle={rootStyle}
			labelColor={labelColor}
			input={
				<Input
					{...textInputProps}
					secureTextEntry={secureTextEntry}
					variant="outline"
					isRequired
					borderColor={error ? 'red.300' : '#E1E5F6'}
					color={'text.900'}
					minH="10"
					mt="1"
					px="4"
					fontFamily="heading"
					fontWeight="400"
					fontSize="14"
					_focus={{ bgColor: 'transparent' }}
					InputLeftElement={
						<ConditionRender.Box renderIf={!!leftIcon} pl="4">
							{leftIcon}
						</ConditionRender.Box>
					}
					InputRightElement={
						<>
							<ConditionRender.Box renderIf={!!rightIcon} pr="4">
								{rightIcon}
							</ConditionRender.Box>

							<ConditionRender.Box renderIf={!!textInputProps.secureTextEntry} pr="2">
								<IconButton
									onPress={toggleSecureTextEntry}
									icon={secureTextEntry ? <EyeIcon /> : <EyeWithLineIcon />}
								/>
							</ConditionRender.Box>
						</>
					}
				/>
			}
		/>
	);
};

export default FormTextInput;
