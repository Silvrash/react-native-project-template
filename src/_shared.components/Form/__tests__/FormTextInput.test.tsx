import { renderWithProviders } from 'src/__mocks__/wrapProviders';
import FormTextInput from '../FormTextInput';

describe('FormTextInput', () => {
	it('should render correctly', () => {
		const { toJSON, getByText } = renderWithProviders(<FormTextInput label="this label" />);

		expect(getByText(/this label/gi)).toBeTruthy();
		expect(toJSON()).toMatchSnapshot();
	});

	it('should render with error ', () => {
		const { toJSON, getByText } = renderWithProviders(
			<FormTextInput label="this label" error="error text" />,
		);

		expect(getByText(/error text/gi)).toBeTruthy();
		expect(toJSON()).toMatchSnapshot();
	});

	it('should render with helper text ', () => {
		const { toJSON, getByText } = renderWithProviders(
			<FormTextInput label="this label" helperText="helper text" />,
		);

		expect(getByText(/helper text/gi)).toBeTruthy();
		expect(toJSON()).toMatchSnapshot();
	});
});
