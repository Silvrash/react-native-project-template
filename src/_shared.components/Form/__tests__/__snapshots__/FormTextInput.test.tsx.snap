// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextInput should render correctly 1`] = `
<View
  dataSet={{}}
  style={
    {
      "minHeight": 36,
    }
  }
>
  <Text
    dataSet={{}}
    style={
      {
        "backgroundColor": undefined,
        "color": "#4C4D6E",
        "fontFamily": "DMSans-Bold",
        "fontSize": 14,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "letterSpacing": 0,
        "lineHeight": 21,
        "marginTop": 8,
        "textDecorationLine": undefined,
      }
    }
  >
    this label
  </Text>
  <View
    dataSet={{}}
    isFocused={false}
    style={
      {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#d4d4d4",
        "borderRadius": 0,
        "borderWidth": 0,
        "flexDirection": "row",
        "minHeight": 40,
        "overflow": "hidden",
      }
    }
  >
    <TextInput
      accessibilityRequired={true}
      accessible={true}
      dataSet={{}}
      editable={true}
      isRequired={true}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      placeholderTextColor="#413f3c80"
      required={true}
      secureTextEntry={false}
      style={
        {
          "backgroundColor": "transparent",
          "color": "#171717",
          "flex": 1,
          "fontFamily": "DMSans-Regular",
          "fontSize": 12,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 0,
          "paddingRight": 12,
          "paddingTop": 8,
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;

exports[`FormTextInput should render with error  1`] = `
<View
  dataSet={{}}
  style={
    {
      "minHeight": 36,
    }
  }
>
  <Text
    dataSet={{}}
    style={
      {
        "backgroundColor": undefined,
        "color": "#4C4D6E",
        "fontFamily": "DMSans-Bold",
        "fontSize": 14,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "letterSpacing": 0,
        "lineHeight": 21,
        "marginTop": 8,
        "textDecorationLine": undefined,
      }
    }
  >
    this label
  </Text>
  <View
    dataSet={{}}
    isFocused={false}
    style={
      {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#d4d4d4",
        "borderRadius": 0,
        "borderWidth": 0,
        "flexDirection": "row",
        "minHeight": 40,
        "overflow": "hidden",
      }
    }
  >
    <TextInput
      accessibilityRequired={true}
      accessible={true}
      dataSet={{}}
      editable={true}
      isRequired={true}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      placeholderTextColor="#413f3c80"
      required={true}
      secureTextEntry={false}
      style={
        {
          "backgroundColor": "transparent",
          "color": "#171717",
          "flex": 1,
          "fontFamily": "DMSans-Regular",
          "fontSize": 12,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 0,
          "paddingRight": 12,
          "paddingTop": 8,
          "width": "100%",
        }
      }
      underlineColorAndroid="#ff0000"
    />
  </View>
  <Text
    dataSet={{}}
    style={
      {
        "backgroundColor": undefined,
        "color": "#f87171",
        "fontFamily": "DMSans-Medium",
        "fontSize": 12,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "letterSpacing": 0,
        "lineHeight": 18,
        "textDecorationLine": undefined,
      }
    }
  >
    error text
  </Text>
</View>
`;

exports[`FormTextInput should render with helper text  1`] = `
<View
  dataSet={{}}
  style={
    {
      "minHeight": 36,
    }
  }
>
  <Text
    dataSet={{}}
    style={
      {
        "backgroundColor": undefined,
        "color": "#4C4D6E",
        "fontFamily": "DMSans-Bold",
        "fontSize": 14,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "letterSpacing": 0,
        "lineHeight": 21,
        "marginTop": 8,
        "textDecorationLine": undefined,
      }
    }
  >
    this label
  </Text>
  <Text
    dataSet={{}}
    style={
      {
        "backgroundColor": undefined,
        "color": "#4C4D6E",
        "fontFamily": "DMSans-Regular",
        "fontSize": 11,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "letterSpacing": 0,
        "lineHeight": 15,
        "textDecorationLine": undefined,
      }
    }
  >
    helper text
  </Text>
  <View
    dataSet={{}}
    isFocused={false}
    style={
      {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#d4d4d4",
        "borderRadius": 0,
        "borderWidth": 0,
        "flexDirection": "row",
        "minHeight": 40,
        "overflow": "hidden",
      }
    }
  >
    <TextInput
      accessibilityRequired={true}
      accessible={true}
      dataSet={{}}
      editable={true}
      isRequired={true}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      placeholderTextColor="#413f3c80"
      required={true}
      secureTextEntry={false}
      style={
        {
          "backgroundColor": "transparent",
          "color": "#171717",
          "flex": 1,
          "fontFamily": "DMSans-Regular",
          "fontSize": 12,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 0,
          "paddingRight": 12,
          "paddingTop": 8,
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;
